### ISSUE TITLE
Create the "Top earning content" key metric widget tile

### ISSUE THREAD

**Issue by jimmymadon** [google/site-kit-wp#6248]
## Feature Description

<img width="384" alt="Screenshot 2022-12-19 at 01 37 53" src="https://user-images.githubusercontent.com/1721319/210290058-d54be5af-0a89-41fb-b798-aef44838d4bd.png">

This key metric widget tile should only appear if users have both Analytics (4) and AdSense modules connected. When there is no `adSenseLinked` account available in Analytics settings (see #8048), this widget should display a CTA asking users to link their AdSense and Analytics (see #8050).

If the `adSenseLinked` account setting is set, the actual content should be displayed.

**NOTE:** This issue must be deployed in the same release as #8051 and #8050.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new Key Metrics Widget, "Top earning content" should be created and available. 
* The Widget will have following attributes:
    * `title`: `Top earning pages`
    * `description`: Pages that generated the most AdSense revenue
    * `slug`: `kmAnalyticsAdSenseTopEarningContent`
    * Metrics: `totalAdRevenue`
    * Dimensions: `pagePath`
    * Value: It should be a table pattern metric tile, with the `pageTitle` showing up as the first column, and totalAdRevenue showing up as the second column (with currency sign).
* The Widget will show top AdSense revenue earning contents (post/page).
* The module should be dependent on both `analytics-4` and `adsense` module. If either module is not connected, the corresponding `Connect{GA4/AdSense}CTATileWidget` should be rendered instead. 
* If both modules are connected but are not linked, a CTA should be displayed instead, asking user to connect their GA4 property with their AdSense account, similar to [8050](https://github.com/google/site-kit-wp/issues/8050).
* The widget should be available for view only users who have access to both Analytics 4 and AdSense module and the configured property and accounts are linked. 

#### Update `assets/js/modules/adsense/components/widgets/TopEarningContentWidget.js`
  * You can borrow good part of the logic from `PopularContentWidget` as starting point
  * [x] Update `reportOptions` with parameters outlined in the AC
  * [x] Leave limit to `3` for consistency with other widgets
  * [x] Check for `select( MODULES_ANALYTICS_4 ).getAdsenseLinked()` and whether the context is view-only dashboard. If the context is not view-only and `adsenseLinked` is false, output early `<AdSenseLinkCTA />` component wrapped with `<Widget>` component
  * [x] When returning metric value in `numFmt` function, pass an object as second parameter `{style: 'currency', currency}`. You can obtain `currency` from `report` object - `report.metadata.currencyCode`

#### Move `AdSenseLinkCTA` component over to the Adsense module
* [x] Move from `assets/js/modules/analytics-4/components/common/AdSenseLinkCTA.js` to the `assets/js/modules/adsense/components/common/AdSenseLinkCTA.js`

#### Add `KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT` to the `assets/js/components/KeyMetrics/key-metrics-widgets.js` 
  * [x] Use `title` and `description` text defined in AC
  * [x] Use `description` text for `infoTooltip`
  * [x] For `displayInList` it should show if following conditions are met:
    * [x] `analytics-4` and `adsense` modules are active
    * [x] Or widget is active, check it using `isKeyMetricActive` selector on `CORE_USER` store
    * [x] Or if in view only context, besides both modules being active,  AdSense should be linked (`getAdSenseLinked()`)

#### Update `assets/js/modules/adsense/index.js`:
* [x] Under `KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT` widget registration, for `isActive` parameter:
  * [x] Besides the current check for `isKeyMetricActive`, check if we are in view only context, and in that case Analytics and AdSense should be linked (`getAdSenseLinked`), otherwise widget should not be displayed for view only users

### Test Coverage

* [ ] Update tests to include new widget in `assets/js/components/KeyMetrics/MetricsSelectionPanel/index.test.js` where applicable
* [x] Add tests and stories for `TopEarningContent`

## QA Brief

* [ ] Enable the `ga4AdSenseIntegration` feature flag
* [ ] View the `TopEarningContentWidget` in Storybook and confirm each state matches designs/AC.
* [ ] Simulate linking your AdSense and Analytics account by updating the `adSenseLinked` setting in the store (this will allow you to QA this ticket before #8049 is merged).
*You can achieve this by running the following the developer console:*
```
googlesitekit.data.dispatch('modules/analytics-4').setAdSenseLinked(true)
```
* [ ] Enable key metric widgets, and enable the "Top earning content" widget. Results from your site, if available, should appear in the widget.

When the `ga4AdSenseIntegration` feature flag is disabled, this KMW tile should not appear.

## Changelog entry

* Add the "Top Earning Content" Key Metric widget tile.

**Comment by jimmymadon** [comment]
Moving this issue to stalled for now as we will have to wait for the GA4 API to support an integration with AdSense.

c.c. @aaemnnosttv

**Comment by bethanylang** [comment]
@tofumatt FYI, we discussed this issue for the Top earning content tile during today's Updates & Planning call. This as one of the tiles we initially had planned for KMW, but the GA4 integration with AdSense wasn't yet ready. Now that we're going to be building the integration, we'd like to add this tile, too.

@aaemnnosttv that we should be able to work on it and get it implemented in the latter part of the AdSense + GA4 epic. I'm going to make a note to myself to move it out of Stalled and into AC once we start moving on engineering for this epic. Let me know if you see any issues with this approach!

**Comment by bethanylang** [comment]
@kuasha420 Just noting that we'd like to include this in the next sprint (starting w/c 12 Feb) so we should probably get it moving to ACR soon. Thanks!

**Comment by ivonac4** [comment]
@kuasha420 can we add an estimate to this task at this point?

**Comment by eugene-manuilov** [comment]
AC ✔️

**Comment by zutigrm** [comment]
Hi @kuasha420 , what content should we use for `infoTooltip`?

**Comment by techanvil** [comment]
Hi @zutigrm, thanks for drafting this IB. The HOC is an interesting idea. However, I don't think we'll need it: we already have the r `AdSenseLinkCTA` rendering defined in https://github.com/google/site-kit-wp/issues/8050, and as you'll see there, we already have the placeholder component `DashboardTopEarningPagesWidgetGA4` for this widget which can be updated.

So, please can you make an appropriate update to this IB.

**Comment by zutigrm** [comment]
Back with you @techanvil for a fresh look, after our discussion in slack

**Comment by techanvil** [comment]
Thanks @zutigrm! Apologies, as discussed on Slack I got mixed up here and you had it right the first time, as this issue is about creating the Key Metrics version of the widget rather than the main dashboard one in the issues I linked to. A few points:

- The HOC remains interesting (if we were to keep it, I would suggest renaming it to `with-conditional-render.js` as this could be applied to any component, not just widgets), but I don't think it's really needed here. Using the HOC means we need to add the additional `LinkAnalyticsAdsenseCTATileWidget` component - it would be more straightforward to follow the approach taken in #8050 and simply render `AdSenseLinkCTA` directly within the widget.
  - AFAICT, `LinkAnalyticsAdsenseCTATileWidget` wouldn't be needed for anything else - we don't need to call `useWidgetStateEffect()` in the way `ConnectGA4CTATileWidget` does, because there is only one AdSense Key Metric tile so we don't need to track the widget state in order to be able to span the widgets.
- Note that we do have the placeholder component for this widget in `assets/js/modules/adsense/components/widgets/TopEarningContentWidget.js`, with a bare bones widget wrapped in the `withActive()` HOCs, so this should be updated rather than created a new widget component.
- We already have the `KM_ANALYTICS_ADSENSE_TOP_EARNING_CONTENT` constant defined.
- Seeing as we're only using `AdSenseLinkCTA` in components in the `adsense` module, we should move `AdSenseLinkCTA` to `adsense` from `analytics` to avoid the cross-module import.
- It seems we should add the check for AdSense being linked for the view-only user to `displayInList()`, otherwise the widget might display in the selection panel but not be present in the widget area.
- I'm not sure if we need to check for shared module access in `isActive()` - the widget will already _not_ be displayed if one of its module dependencies is no longer shared. Please can you check and confirm this?

**Comment by zutigrm** [comment]
Thank you @techanvil for the feedback!

 > I'm not sure if we need to check for shared module access in isActive() - the widget will already not be displayed if one of its module dependencies is no longer shared. Please can you check and confirm this?
 
 Yes, you are correct, thanks for the pointer
 
 IB Updated

**Comment by techanvil** [comment]
Thanks @zutigrm! I think we should still include the check for `getAdSenseLinked()` in `isActive()` though, shouldn't we?

**Comment by zutigrm** [comment]
@techanvil ah I deleted whole piece instead of just the has access to shareable modules part. Edited

**Comment by techanvil** [comment]
Cool, thanks @zutigrm. IB LGTM! :white_check_mark:

**Comment by mohitwp** [comment]
@wpdarren As mentioned in slack are you able to link AdSense and Analytics using oi.ie or any of your live site and not getting prompt if analytics and AdSense already links. Like as I mentioned in slack happening with oi.ie site ?

**Comment by techanvil** [comment]
I believe this issue depends on https://github.com/google/site-kit-wp/issues/8049 to be fully E2E testable, otherwise we'd need to manually set the `adSenseLinked` setting to test it on the dashboard.

However, it does need a followup PR as per https://github.com/google/site-kit-wp/pull/8249#pullrequestreview-1883267070. I'm assigning back to @benbowler in Execution for the fix.

**Comment by benbowler** [comment]
@tofumatt @zutigrm I have pushed a few updates to the PR #8257 and have tested it myself and I'm able to view it successfully when manually updating the `adSenseLinked` state in the store. I've updated the QAB with instructions on how to QA this without the need for #8049 to be merged.

Back into CR with you @tofumatt

**Comment by mohitwp** [comment]
### QA Update ✅ 


- Tested on dev environment.
- Verified when `ga4AdSenseIntegration` feature flag is enabled and both AdSense and Analytcis are connected then and only KM 'Top earning pages' widgets showing under the KM widgets listing and on selecting it prompt implemented under #8050  is appearing.
- Verified when `ga4AdSenseIntegration` feature flag is not enabled then KM 'Top earning pages' widgets showing under the KM widgets listing.
- Verfied widgets Title, Tooltip and slug details as per AC.
- Verified that on View only dashboard link not appear for the pages.
- Verified that if Analytics and AdSense are not linked then Key metrics `Top earning page` widgets is not showing under widgets listing on 'View - Only Dashboard'.
- Verified that if only Analytics is shared then then Key metrics `Top earning page` widgets is not showing under widgets listing on 'View - Only Dashboard'.
- Verfiied that if Only AdSense is shared then Key metrics tiles are not showing.
- Verified that if Analytics and AdSense are both connected. Shared and linked together then  `Top earning page` widgets is showing under widgets listing on 'View - Only Dashboard' and user is able to select this KM widget.
- 

<details>

**Tool Tip Title and slug-**

![image](https://github.com/google/site-kit-wp/assets/94359491/a729027e-1a50-43ea-9bb4-4fd8e5df433e)

![image](https://github.com/google/site-kit-wp/assets/94359491/0893194d-4ce4-4c9e-b514-1db4874be1b3)

**When both Analytics and AdSense are connected -**

![image](https://github.com/google/site-kit-wp/assets/94359491/eddee7a8-c4f0-423a-89ed-af07f201a138)

**If Analytics is not connected -**

![image](https://github.com/google/site-kit-wp/assets/94359491/afaedd43-23bf-480c-a0a7-8ecf65f75ecb)

**If AdSense is not connected -**

![image](https://github.com/google/site-kit-wp/assets/94359491/3cb2803b-897a-44be-af89-30781aa23cd7)


`Only Analytics is shared (View Only Dashboard')`

![image](https://github.com/google/site-kit-wp/assets/94359491/3baae009-6b04-4c23-a722-13a921a1d8c0)


When both Analytics and Adsense are connected and linked. Also, shared with non admin users ('View Only Dashboard'

![image](https://github.com/google/site-kit-wp/assets/94359491/74814ed8-2352-412d-9317-22cd650e81c2)

![image](https://github.com/google/site-kit-wp/assets/94359491/13e29d94-9c82-4e44-a3d1-81fdbd73238a)

</details>

**Comment by mohitwp** [comment]
**_Note : Putting note here to test again when https://github.com/google/site-kit-wp/issues/8049 get merge._**
